Support streaming from multiple pipeline components

mpangrazzi opened the pull request:
Should close #144 and #133.

Currently in Hayhooks we support streaming only considering the last streaming-capable component.

With this PR, we want to support all streaming-capable components in a pipeline. This can enable support from some specific use cases.

Example where we have a pipeline with 2 LLM-based components:

![multi_stream](https://github.com/user-attachments/assets/5d8519eb-5b3e-4822-b1ba-869dc273d9c1)

Full example: https://github.com/deepset-ai/hayhooks/tree/multi-component-streaming/examples/pipeline_wrappers/multi_llm_streaming

Sample pipeline and `PipelineWrapper`: https://github.com/deepset-ai/hayhooks/blob/multi-component-streaming/examples/pipeline_wrappers/multi_llm_streaming/pipeline_wrapper.py

**NOTE**: I was wondering if to keep the "streaming all components" a default thing, or if it worth make an user to choice between "stream all capable components" and "stream only the last capable one".

anakin87 reviewed (commented):
> I was wondering if to keep the "streaming all components" a default thing, or if it worth make an user to choice between "stream all capable components" and "stream only the last capable one".

I'm not sure... Maybe I would let the user choose.

On the other hand, if streaming isn't needed, the user could simply avoid providing a `streaming_callback`.

Curious to hear @sjrl's thoughts as well...

mpangrazzi commented:
@sjrl 

> Turning it on for all by default could cause problems in edge cases where two LLMs in two branches are running simultaneously (like could happen when using AsyncPipeline).

Speaking with @tstadel this usually doesn't happen due to XOR branches (so one generator at the time is streaming). But assuming it will happen, we may get the source component from the streaming chunk and "fix" the output stream accordingly.

@anakin87 Probably a better solution would be:

- Stream only the _last_ capable component by default
- Accept a param to enable streaming on all capable components (assuming they will do it serially)
- Accept a param to exclude some specific component from streaming (e.g. by name)

This should cover all use cases. For YAML pipelines, same streaming configuration may be read from a `streaming_config` field (as for `inputs` / `outputs`). 

WDYT?

mpangrazzi commented on examples/pipeline_wrappers/multi_llm_streaming/pipeline_wrapper.py:
```diff
@@ -0,0 +1,128 @@
+
+        # Connect the components
+        self.pipeline.connect("prompt_builder_1.prompt", "llm_1.messages")
+        self.pipeline.connect("llm_1.replies", "prompt_builder_2.template")
```
BTW, I'll fix it by adding a custom component which extract the reply from `llm_1` and pass it to `prompt_builder_2`.

sjrl replied:
> BTW, I'll fix it by adding a custom component which extract the reply from `llm_1` and pass it to `prompt_builder_2`.

I don't think you need to a custom component to do that. You should be able to directly pass the `replies` to the `PromptBuilder` and manipulate what you need in the jinja2 template. E.g. `{{replies[0].text}}` would then be  `self.pipeline.connect("llm_1.replies", "prompt_builder_2.replies")`

anakin87 commented on examples/pipeline_wrappers/multi_llm_streaming/README.md:
```diff
@@ -0,0 +1,119 @@
+- **Streaming**: Serial execution with automatic callback management for configured components
+- **Transition Detection**: Uses `StreamingChunk.component_info.name` to detect when LLM 2 starts
+- **Visual Separator**: Injects a `StreamingChunk` between LLM outputs
+- **Error Handling**: Stream terminates gracefully if any component fails
```
I think that this README can be shortened, by removing some AI-generated content

anakin87 commented on tests/test_files/yaml/multi_llm_streaming_all_pipeline.yml:
```diff
@@ -0,0 +1,64 @@
+      model: gpt-4o-mini
+      streaming_callback: null
+      system_prompt: null
+    type: haystack.components.generators.openai.OpenAIGenerator
```
Since at some point, we'll remove Generators and use Chat Generators instead, it's better not to introduce new code based on them

mpangrazzi replied:
True, but WDYT about open an issue and change all `OpenAIGenerator` to `OpenAIChatGenerator` in a separate PR?

anakin87 reviewed (commented):
Good work!

I've left some minor comments.

mpangrazzi commented:
Ok so I've did the following things:

- Update `streaming_components` format from boolean dict `{"llm_1": True, "llm_2": False}` to (white) list `["llm_1"]`, for both YAML and method params
- Fixed an old issue which doesn't make the streaming able to work in concurrency due to override of component's `streaming_callback`. Now it's passed as a run arg. Added also a concurrency integration test [here](https://github.com/deepset-ai/hayhooks/blob/multi-component-streaming/tests/test_it_concurrent_streaming.py) - cc @sjrl 
- Fixed some minor type issues due to `Literal` usage
- Updated docs, examples and tests

cc @tstadel @anakin87

sjrl commented on docs/concepts/pipeline-wrapper.md:
```diff
@@ -176,6 +176,179 @@ async def run_chat_completion_async(self, model: str, messages: List[dict], body
+
+connections:
+  - sender: llm_1.replies
+    receiver: llm_2.prompt
```
Somewhat odd of an example since this connection would fail. `llm_2.prompt` is of type string and `llm_1.replies` is of type `list[str]` since we are using the `OpenAIGenerator` and not the chat version.
